### PR TITLE
nvs: implement Non-volatile Storage cache

### PR DESCRIPTION
**Pull request by Damian-Nordic**
It's an attempt to solve performance issues which we observe when using Settings over NVS: https://github.com/zephyrproject-rtos/zephyr/issues/45591.

The NVS data lookup time grows linearly with the number of allocation table entries to walk through, meaning that if
some data pair in the NVS changes frequently, access to other data pairs that change rarely can take a lot of time.

It is particularly visible when the NVS is used as the settings backend since the backend needs to perform multiple NVS reads to find a requested key.

Implement a simple cache that stores an address of the most recent ATE for all NVS IDs that fall into the given cache
position. CRC8/16 is used as a hash function used to distribute NVS IDs across the cache entries.

The cache entries are only invalidated when an NVS sector is erased as part of the garbage collector task.

Signed-off-by: Damian Krolik <damian.krolik@nordicsemi.no>

### CODE REVIEW

File: tests/subsys/fs/nvs/boards/native_posix.overlay [hunk 1]
```diff
@@ -0,0 +1,9 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&flash0 {
+	erase-block-size = <0x400>;
+};
```
Comment (Laczen) [inline tests/subsys/fs/nvs/boards/native_posix.overlay:9]: Is this needed for the PR ?

File: tests/subsys/fs/nvs/src/main.c [hunk 2]
```diff
@@ -8,11 +8,11 @@
...
+ * Test should be run on qemu_x86 or native_posix target.
  */
 
-#ifndef CONFIG_BOARD_QEMU_X86
-#error "Run on qemu_x86 only"
+#if !defined(CONFIG_BOARD_QEMU_X86) && !defined(CONFIG_BOARD_NATIVE_POSIX)
+#error "Run on qemu_x86 or native_posix only"
 #endif
 
 #include <stdio.h>
```
Comment (Laczen) [inline tests/subsys/fs/nvs/src/main.c:18]: Are these changes needed for the PR ?

File: subsys/fs/nvs/nvs.c [hunk 3]
```diff
@@ -16,6 +16,75 @@
 #include <zephyr/logging/log.h>
 LOG_MODULE_REGISTER(fs_nvs, CONFIG_NVS_LOG_LEVEL);
 
+static int nvs_prev_ate(struct nvs_fs *fs, uint32_t *addr, struct nvs_ate *ate);
+static int nvs_ate_valid(struct nvs_fs *fs, const struct nvs_ate *entry);
+
+#ifdef CONFIG_NVS_LOOKUP_CACHE
+
+#define NVS_LOOKUP_CACHE_EMPTY 0
+
```
Comment (Laczen) [inline subsys/fs/nvs/nvs.c:25]: `NVS_LOOKUP_CACHE_NO_ADDRESS` ? In theory a ATE could be found at address 0, in practice this case doesn't matter because it would be a delete. Maybe this could be added as a remark for future questions.

**Review by Laczen — COMMENTED** [review]
In general I see no objection against the proposed idea. Probably it would be good to add some test for the caching.

File: tests/subsys/fs/nvs/boards/native_posix.overlay (hunk 1 above)
Comment (Damian-Nordic) [reply tests/subsys/fs/nvs/boards/native_posix.overlay:9]: The test assumes such a sector size, and otherwise works fine for native_posix platform. I used the native_posix platform while developing the cache and thought it would be nice to make the test work on that platform. I can extract it to a separate PR if you would prefer so.

File: subsys/fs/nvs/nvs.c (hunk 3 above)
Comment (Damian-Nordic) [reply subsys/fs/nvs/nvs.c]: Changed both the constant name and value.

**Comment by Damian-Nordic** [comment]
@Laczen I added unit tests for the change. They helped me discover that the previous code was missing skipping 0xffff ids. Please take another look.

**Comment by Laczen** [comment]
> @Laczen I added unit tests for the change. They helped me discover that the previous code was missing skipping 0xffff ids. Please take another look.

Nice, good that you added the unit tests. For me the proposed PR is ok, to avoid having to revisit the PR I will add the approval after maintainers have reviewed.

@nvlsianpu, @de-nordic, could you have a look?

**Review by de-nordic — DISMISSED** [review]
Looks OK.

**Review by de-nordic — APPROVED** [review]

**Review by Laczen — APPROVED** [review]

**Review by nvlsianpu — APPROVED** [review]

**Review by pdgendt — APPROVED** [review]

File: subsys/fs/nvs/nvs.c [hunk 4]
```diff
@@ -1051,7 +1131,16 @@ ssize_t nvs_read_hist(struct nvs_fs *fs, uint16_t id, void *data, size_t len,
...
+	if (wlk_addr == NVS_LOOKUP_CACHE_EMPTY) {
+		rc = -ENOENT;
+		goto err;
+	}
+#else
 	wlk_addr = fs->ate_wra;
+#endif
 	rd_addr = wlk_addr;
 
 	while (cnt_his <= cnt) {
```
[the remark this answers is not in the document]
Comment (aurel32) [reply subsys/fs/nvs/nvs.c:1152]: From what i understand, multiple IDs can use the same cache position (especially with small caches). How this case is handled?

[the remark this answers is not in the document]
Comment (Damian-Nordic) [reply subsys/fs/nvs/nvs.c:1152]: If two entries, A and B, share the same cache position, the cache will hold the location of the more recent of the two. Thus, the cache may speed the lookup up even in the case of the cache collision.

[the remark this answers is not in the document]
Comment (aurel32) [reply subsys/fs/nvs/nvs.c:1152]: Ok, understood, thanks for the explanations
